store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java

```diff
@@ -261,24 +258,96 @@ private void recover(long maxPhyOffsetOfConsumeQueue) {
             if (mappedFile != null) {
                 disableDeleteDledger();
             }
-            long maxPhyOffset = dLedgerFileList.getMaxWrotePosition();
+            List<MmapFile> mmapFiles = dLedgerFileList.getMappedFiles();
+            int index = mmapFiles.size() - 1;
+            MmapFile mmapFile = null;
+            for (; index >= 0; index--) {
+                mmapFile = mmapFiles.get(index);
+                if (isMappedFileMatchedRecover(mmapFile)) {
+                    log.info("Recover from this dleger mapped file " + mmapFile.getFileName());
+                    break;
+                }
+            }
+            if (index < 0) {
+                index = 0;
+                mmapFile = mmapFiles.get(index);
+            }
+
+            ByteBuffer byteBuffer = mmapFile.sliceByteBuffer();
+            long processOffset = mmapFile.getFileFromOffset();
+            long mappedFileOffset = 0;
+            while (true) {
+                DispatchRequest dispatchRequest = this.checkMessageAndReturnSize(byteBuffer, checkCRCOnRecover);
+                int size = dispatchRequest.getMsgSize();
```

Comment:
This size does not contain dledger header, causing the process offset  to be wrong. Maybe should add DLedgerEntry.BODY_OFFSET.